src/decoder/lattice-faster-decoder-combine-bucketqueue.cc

```diff
@@ -0,0 +1,1123 @@
+                                BaseFloat cost_scale) :
+    cost_scale_(cost_scale) {
+  // NOTE: we reserve plenty of elements to avoid expensive reallocations
+  // later on. Normally, the size is a little bigger than (adaptive_beam +
+  // 5) * cost_scale.
+  int32 bucket_size = 100;
+  buckets_.resize(bucket_size);
+  bucket_storage_begin_ = std::floor((best_cost_estimate - 15) * cost_scale_);
+  first_occupied_vec_index_ = bucket_size;
+}
+
+template<typename Token>
+void BucketQueue<Token>::Push(Token *tok) {
+  int32 bucket_index = std::floor(tok->tot_cost * cost_scale_);
+  size_t vec_index = static_cast<size_t>(bucket_index - bucket_storage_begin_);
+  if (vec_index >= buckets_.size()) {
+    int32 margin = 10;  // a margin which is used to reduce re-allocate
+                        // space frequently
+    // A cast from unsigned to signed type does not generate a machine-code
+    // instruction
+    if (static_cast<int32>(vec_index) > 0) {
+      KALDI_WARN << "Have to reallocate the BucketQueue. Maybe need to reserve"
+                 << " more elements in constructor. Push back.";
+      buckets_.resize(static_cast<int32>(vec_index) + margin);
+    } else {  // less than 0
+       KALDI_WARN << "Have to reallocate the BucketQueue. Maybe need to reserve"
+                 << " more elements in constructor. Push front.";
+       int32 increase_size = - static_cast<int32>(vec_index) + margin;
+       buckets_.resize(buckets_.size() + increase_size);
+       // translation
+       for (size_t i = buckets_.size() - 1; i >= increase_size; i--) {
+         buckets_[i].swap(buckets_[i - increase_size]);
+       }
+       bucket_storage_begin_ = bucket_storage_begin_ - increase_size;
+       vec_index = static_cast<int32>(vec_index) + increase_size;
+       first_occupied_vec_index_ = vec_index;
+    }
+  }
+  tok->in_queue = true;
+  buckets_[vec_index].push_back(tok);
+  if (vec_index < first_occupied_vec_index_)
+    first_occupied_vec_index_ = vec_index;
+}
+
+template<typename Token>
+Token* BucketQueue<Token>::Pop() {
+  int32 vec_index = first_occupied_vec_index_;
+  while (vec_index < buckets_.size()) {
+    Token* tok = buckets_[vec_index].back();
+    // Remove the best token
+    buckets_[vec_index].pop_back();
+
+    if (buckets_[vec_index].empty()) { // This bucket is empty. Update vec_index
+      int32 next_vec_index = vec_index + 1;
+      for (; next_vec_index < buckets_.size(); next_vec_index++) {
+        if (!buckets_[next_vec_index].empty()) break;
+      }
+      vec_index = next_vec_index;
+      first_occupied_vec_index_ = vec_index;
+    }
+
+    if (tok->in_queue) {  // This is a effective token
+      tok->in_queue = false;
+      return tok;
+    }
+  }
+  return NULL;
+}
+
+template<typename Token>
+void BucketQueue<Token>::Clear() {
+  for (size_t i = 0; i < buckets_.size(); i++) {
+    buckets_[i].clear();
+  }
+  first_occupied_vec_index_ = buckets_.size();
```

Comment:
obviously if you make the changes to first_occupied_vec_index_ I mentioned above, this line would change, as it has to be < buckets_.size().